### PR TITLE
openblas:build_lapack should use gfortran from CCI

### PR DESCRIPTION
**Pull request by ggulgulia**
Library name  **openblas packages**

* Openblas relies on system gfortran compiler for building with the option `build_lapack`, which on many systems may not available out of the box
* This PR changes openblas recipe to use gfortran from conan-center-index 
* Additionally update the recipe to use conan V2
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/openblas/all/conanfile.py [hunk 1]
```diff
@@ -50,6 +50,10 @@ def configure(self):
         if self.options.shared:
             del self.options.fPIC
 
+    def requirements(self):
```
Comment (uilianries) [inline recipes/openblas/all/conanfile.py:53]: I believe gfortran is a build requirement, does not it? You want fortran to build lapack only.

Comment (ggulgulia) [reply recipes/openblas/all/conanfile.py:53]: that is true. I should use build_requires I suppose there.

**Comment by ggulgulia** [comment]
I'm still not able to build openblas with lapack after the modifications. When I run the command 
```
conan create path/to/openblas/conanfile.py 0.3.17@user/testing -o openblas:build_lapack=True
```

I get an error indicating that fortran compiler is not found
```
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - NOTFOUND
CMake Error at source_subfolder/cmake/f_check.cmake:29 (message):
  No Fortran compiler found.  Cannot build with LAPACK.
Call Stack (most recent call first):
  source_subfolder/cmake/prebuild.cmake:89 (include)
  source_subfolder/cmake/system.cmake:148 (include)
  source_subfolder/CMakeLists.txt:73 (include)


-- Configuring incomplete, errors occurred!
See also "/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9/build_subfolder/CMakeFiles/CMakeOutput.log".
See also "/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9/build_subfolder/CMakeFiles/CMakeError.log".
```

From our previous discussion on simbody MR, I conclude that we cannot change the option `build_lapack=True` in the recipe to test if this works on conan-center-index CI

## Build Log

<details><summary>Click to expand log</summary>

```
ggulgulia:~/workspace/conan-center-index/recipes/openblas/all$ conan create . 0.3.17@cc/testing -o openblas:build_lapack=True
WARN: Hook 'ar/set_compiler.py' not found in /home/g.gulgulia/.conan/hooks folder. Please remove hook from conan.conf or include it inside the hooks folder.
[HOOK - conan-center.py] pre_export(): [DEPRECATED GLOBAL CPPSTD (KB-H001)] OK
[HOOK - conan-center.py] pre_export(): [REFERENCE LOWERCASE (KB-H002)] OK
[HOOK - conan-center.py] pre_export(): [RECIPE METADATA (KB-H003)] OK
[HOOK - conan-center.py] pre_export(): [HEADER_ONLY, NO COPY SOURCE (KB-H005)] OK
[HOOK - conan-center.py] pre_export(): [FPIC OPTION (KB-H006)] OK
[HOOK - conan-center.py] pre_export(): [VERSION RANGES (KB-H008)] OK
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] Total recipe size: 8.279296875 KB
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] OK
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE FOLDER (KB-H024)] OK
[HOOK - conan-center.py] pre_export(): [META LINES (KB-H025)] OK
[HOOK - conan-center.py] pre_export(): [CONAN CENTER INDEX URL (KB-H027)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE MINIMUM VERSION (KB-H028)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - RUN ENVIRONMENT (KB-H029)] OK
[HOOK - conan-center.py] pre_export(): [SYSTEM REQUIREMENTS (KB-H032)] OK
[HOOK - conan-center.py] pre_export(): [CONANDATA.YML FORMAT (KB-H030)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - NO IMPORTS() (KB-H034)] OK
[HOOK - conan-center.py] pre_export(): [NO AUTHOR (KB-H037)] OK
[HOOK - conan-center.py] pre_export(): [NOT ALLOWED ATTRIBUTES (KB-H039)] OK
[HOOK - conan-center.py] pre_export(): [NO TARGET NAME (KB-H040)] OK
[HOOK - conan-center.py] pre_export(): [NO FINAL ENDLINE (KB-H041)] OK
[HOOK - conan-center.py] pre_export(): [NO REQUIRES.ADD() (KB-H044)] OK
[HOOK - conan-center.py] pre_export(): [DELETE OPTIONS (KB-H045)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE VERBOSE MAKEFILE (KB-H046)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE VERSION REQUIRED (KB-H048)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE WINDOWS EXPORT ALL SYMBOLS (KB-H049)] OK
[HOOK - conan-center.py] pre_export(): [DEFAULT OPTIONS AS DICTIONARY (KB-H051)] OK
[HOOK - conan-center.py] pre_export(): [CONFIG.YML HAS NEW VERSION (KB-H052)] OK
[HOOK - conan-center.py] pre_export(): [PRIVATE IMPORTS (KB-H053)] OK
[HOOK - conan-center.py] pre_export(): [SINGLE REQUIRES (KB-H055)] OK
[HOOK - conan-center.py] pre_export(): [TOOLS RENAME (KB-H057)] OK
[HOOK - conan-center.py] pre_export(): [ILLEGAL CHARACTERS (KB-H058)] OK
[HOOK - conan-center.py] pre_export(): [CLASS NAME (KB-H059)] OK
[HOOK - conan-center.py] pre_export(): [NO CRLF (KB-H060)] OK
[HOOK - conan-center.py] pre_export(): [NO BUILD SYSTEM FUNCTIONS (KB-H061)] OK
[HOOK - conan-center.py] pre_export(): [TOOLS CROSS BUILDING (KB-H062)] OK
[HOOK - conan-center.py] pre_export(): [INVALID TOPICS (KB-H064)] OK
[HOOK - conan-center.py] pre_export(): [NO REQUIRED_CONAN_VERSION (KB-H065)] OK
[HOOK - conan-center.py] pre_export(): [TEST_TYPE MANAGEMENT (KB-H068)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - NO DEFAULT OPTIONS (KB-H069)] OK
[HOOK - conan-center.py] pre_export(): [MANDATORY SETTINGS (KB-H070)] OK
Exporting package recipe
openblas/0.3.17@cc/testing exports: File 'conandata.yml' found. Exporting it...
openblas/0.3.17@cc/testing exports: Copied 1 '.yml' file: conandata.yml
openblas/0.3.17@cc/testing: Calling export_sources()
openblas/0.3.17@cc/testing export_sources() method: Copied 1 '.txt' file: CMakeLists.txt
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] Saving conandata.yml: {'sources': {'0.3.17': {'url': 'https://github.com/xianyi/OpenBLAS/archive/v0.3.17.tar.gz', 'sha256': 'df2934fa33d04fd84d839ca698280df55c690c86a5a1133b3f7266fce1de279f'}}}
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] New conandata.yml contents: sources:
  0.3.17:
    sha256: df2934fa33d04fd84d839ca698280df55c690c86a5a1133b3f7266fce1de279f
    url: https://github.com/xianyi/OpenBLAS/archive/v0.3.17.tar.gz

[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] OK
[HOOK - conan-center.py] post_export(): [DEFAULT SHARED OPTION VALUE (KB-H050)] OK
[HOOK - conan-center.py] post_export(): [LICENSE PUBLIC DOMAIN (KB-H056)] OK
openblas/0.3.17@cc/testing: A new conanfile.py version was exported
openblas/0.3.17@cc/testing: Folder: /home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/export
openblas/0.3.17@cc/testing: Using the exported files summary hash as the recipe revision: b74b31d1afc5c8f1beb2cad940b31ea2 
openblas/0.3.17@cc/testing: Package recipe modified in export, forcing source folder removal
openblas/0.3.17@cc/testing: Use the --keep-source, -k option to skip it
openblas/0.3.17@cc/testing: Exported revision: b74b31d1afc5c8f1beb2cad940b31ea2
Configuration:
[settings]
arch=x86_64
arch.march=None
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=7
os=Linux
os_build=Linux
[options]
openblas:build_lapack=True
[build_requires]
[env]

openblas/0.3.17@cc/testing: Forced build from source
openblas/0.3.17@cc/testing (test package): Installing package
Requirements
    openblas/0.3.17@cc/testing from local cache - Cache
Packages
    openblas/0.3.17@cc/testing:fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9 - Build

Installing (downloading, building) binaries...
[HOOK - conan-center.py] pre_source(): [IMMUTABLE SOURCES (KB-H010)] OK
openblas/0.3.17@cc/testing: Configuring sources in /home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/source
Downloading v0.3.17.tar.gz completed [12219.76k]                                         

[HOOK - conan-center.py] post_source(): [LIBCXX MANAGEMENT (KB-H011)] OK
[HOOK - conan-center.py] post_source(): [CPPSTD MANAGEMENT (KB-H022)] OK
[HOOK - conan-center.py] post_source(): [SHORT_PATHS USAGE (KB-H066)] OK
openblas/0.3.17@cc/testing: Copying sources to build folder
openblas/0.3.17@cc/testing: Building your package in /home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9
openblas/0.3.17@cc/testing: Generator cmake created conanbuildinfo.cmake
openblas/0.3.17@cc/testing: Aggregating env generators
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK. 'fPIC' option found and apparently well managed
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
openblas/0.3.17@cc/testing: Calling build()
openblas/0.3.17@cc/testing: WARN: Building with lapack support requires a Fortran compiler.
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Adjusting fPIC flag (ON)
-- Conan: Compiler GCC>=5, checking major version 7
-- Conan: Checking correct version: 7
-- Conan: C++ stdlib: libstdc++11
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
CMake Warning at source_subfolder/CMakeLists.txt:70 (message):
  CMake support is experimental.  It does not yet support all build options
  and may not produce the same Makefiles that OpenBLAS ships with.


-- Looking for stdatomic.h
-- Looking for stdatomic.h - found
-- GEMM multithread threshold set to 4.
-- Multi-threading enabled with 12 threads.
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - NOTFOUND
CMake Error at source_subfolder/cmake/f_check.cmake:29 (message):
  No Fortran compiler found.  Cannot build with LAPACK.
Call Stack (most recent call first):
  source_subfolder/cmake/prebuild.cmake:89 (include)
  source_subfolder/cmake/system.cmake:148 (include)
  source_subfolder/CMakeLists.txt:73 (include)


-- Configuring incomplete, errors occurred!
See also "/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9/build_subfolder/CMakeFiles/CMakeOutput.log".
See also "/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9/build_subfolder/CMakeFiles/CMakeError.log".
openblas/0.3.17@cc/testing: 
openblas/0.3.17@cc/testing: ERROR: Package 'fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9' build failed
openblas/0.3.17@cc/testing: WARN: Build folder /home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9
ERROR: openblas/0.3.17@cc/testing: Error in build() method, line 116
        cmake = self._configure_cmake()
while calling '_configure_cmake', line 90
        cmake.configure(build_folder=self._build_subfolder)
        ConanException: Error 1 while executing cd '/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9/build_subfolder' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCONAN_IN_LOCAL_CACHE="ON" -DCONAN_COMPILER="gcc" -DCONAN_COMPILER_VERSION="7" -DCONAN_CXX_FLAGS="-m64" -DCONAN_SHARED_LINKER_FLAGS="-m64" -DCONAN_C_FLAGS="-m64" -DCONAN_LIBCXX="libstdc++11" -DBUILD_SHARED_LIBS="OFF" -DCMAKE_INSTALL_PREFIX="/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/package/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCONAN_CMAKE_POSITION_INDEPENDENT_CODE="ON" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -DNOFORTRAN="False" -DBUILD_WITHOUT_LAPACK="False" -DDYNAMIC_ARCH="False" -DUSE_THREAD="True" -DUSE_LOCKING="False" -DMSVC_STATIC_CRT="False" -DANDROID="True" -Wno-dev '/home/g.gulgulia/.conan/data/openblas/0.3.17/cc/testing/build/fc0e9c0c36235c3c4235c08d830f7b48f44ad4e9'
```

</details>

File: recipes/openblas/all/conanfile.py [hunk 2]
```diff
@@ -50,6 +50,10 @@ def configure(self):
         if self.options.shared:
             del self.options.fPIC
 
+    def build_requirements(self):
+        if self.options.build_lapack:
+            self.requires("gfortran/10.2")
```
Comment (uilianries) [inline recipes/openblas/all/conanfile.py:55]:
```suggestion
            self.tool_requires("gfortran/10.2")
```
self.requires should be used only for def requirements()

**Comment by prince-chrismc** [comment]
How does the project test for the compiler?

```
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - NOTFOUND
CMake Error at source_subfolder/cmake/f_check.cmake:29 (message):
  No Fortran compiler found.  Cannot build with LAPACK.
Call Stack (most recent call first):
  source_subfolder/cmake/prebuild.cmake:89 (include)
  source_subfolder/cmake/system.cmake:148 (include)
  source_subfolder/CMakeLists.txt:73 (include)
```

I would assume there's some hint or path check and we need to give the CMake the location of the compiler?

**Comment by jwillikers** [comment]
> How does the project test for the compiler?
> 
> ```
> -- Looking for a Fortran compiler
> -- Looking for a Fortran compiler - NOTFOUND
> CMake Error at source_subfolder/cmake/f_check.cmake:29 (message):
>   No Fortran compiler found.  Cannot build with LAPACK.
> Call Stack (most recent call first):
>   source_subfolder/cmake/prebuild.cmake:89 (include)
>   source_subfolder/cmake/system.cmake:148 (include)
>   source_subfolder/CMakeLists.txt:73 (include)
> ```
> 
> I would assume there's some hint or path check and we need to give the CMake the location of the compiler?

Likely, `gfortran` needs to be installed. Otherwise, CMake documents setting the Fortran compiler [here](https://cmake.org/cmake/help/latest/envvar/FC.html). Use the `FC` environment variable or the `CMAKE_Fortran_COMPILER` cache variable.

**Comment by prince-chrismc** [comment]
That's not being set https://github.com/conan-io/conan-center-index/blob/4bf114eefecf85163341d07233fa94b6186ae97c/recipes/gfortran/all/conanfile.py#L74

**Comment by jwillikers** [comment]
> That's not being set
> 
> https://github.com/conan-io/conan-center-index/blob/4bf114eefecf85163341d07233fa94b6186ae97c/recipes/gfortran/all/conanfile.py#L74

Wow, I had no idea we had a `gfortran` Conan package. Oh... it's a binary package. Bummer. Yeah, setting the environment variable should work for Autotools, too, per the documentation [here](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Fortran-Compiler.html#index-FC-972).

**Comment by uilianries** [comment]
@jwillikers There is a long history behind gfortran package. It was totally broken for many months, then we converted to a pre-built tool recipe.

**Comment by samuel-emrys** [comment]
I think the gfortran package is still broken as per https://github.com/conan-io/conan-center-index/issues/14120, even after making the necessary modifications to export `FC` to the build environment. I've had limited success on linux building lapack using the gfortran executable build with the gcc as updated in https://github.com/conan-io/conan-center-index/pull/13896 but this still won't be useful for modifying the default option until the `gcc` recipe supports both windows and macos as well.

**Pull request closed without merging**